lib/pbench/server/api/resources/upload_api.py

```diff
@@ -236,9 +238,24 @@ def _put(self, args: ApiParams, request: Request, context: ApiContext) -> Respon
                     f"'Content-Length' {content_length} must be greater than 0",
                 )
 
-            tar_full_path = self.temporary / filename
-            md5_full_path = self.temporary / f"{filename}.md5"
-            dataset_name = Dataset.stem(tar_full_path)
+            dataset_name = Dataset.stem(filename)
+
+            # NOTE: we isolate each uploaded tarball into a private MD5-based
+            # subdirectory in order to retain the original tarball stem name
+            # for the cache manager while giving us protection against multiple
+            # tarballs with the same name. (A duplicate MD5 will have already
+            # failed, so that's not a concern.)
+            try:
+                tmp_dir = self.temporary / md5sum
+                tmp_dir.mkdir()
+            except FileExistsError:
+                raise CleanupTime(
+                    HTTPStatus.CONFLICT,
+                    "Temporary upload directory already exists",
+                )
+            upload_dir = tmp_dir
```

Comment:
Why have both `upload_dir` and `tmp_dir`?  It seems like either one would suffice.

Reply:
It was a journey ... but you're right, they're now redundant and it does look a bit odd.